sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InsertIntoHadoopFsRelation.scala

```diff
@@ -160,6 +160,9 @@ private[sql] case class InsertIntoHadoopFsRelation(
       logInfo("Skipping insertion into a relation that already exists.")
     }
 
+    // Invalidate the cache.
+    sqlContext.cacheManager.invalidateCache(LogicalRelation(relation))
```

Comment:
This line should be moved right after the `relation.refresh()` call above. Since we don't want to invalidate the cache when

1. no insertion needs to be performed, or
2. the write job fails